### PR TITLE
Document conversion functions

### PR DESCRIPTION
**Pull request by gautierronan**
- Add docstrings for all QArray conversion functions
- Merge `qarrays/utils.py` and `qarrays/type_conversion.py` into a single file (`qarrays/utils.py`) for single table in documentation API index
- Add missing `dq.to_numpy` function
- Minor fixes in these functions

Suggest reviewing commit by commit.

### CODE REVIEW

File: dynamiqs/time_array.py [hunk 1]
```diff
@@ -206,13 +206,13 @@ def timecallable(
...
+        QArray: shape=(2, 2), dims=(2,), dtype=complex64, layout=dense
+        [[0.5+0.j 0. +0.j]
+         [0. +0.j 0.5+0.j]]
         >>> H(1.0)
-        QArray: shape=(2, 2), dims=(2,), dtype=float32, layout=dense
-        [[1. 0.]
-         [0. 0.]]
+        QArray: shape=(2, 2), dims=(2,), dtype=complex64, layout=dense
+        [[1.+0.j 0.+0.j]
+         [0.+0.j 0.+0.j]]
```
Comment (pierreguilmin) [inline dynamiqs/time_array.py:215]: This is quite annoying. Is there a simple workaround? (we can merge the PR and fix it later)

Comment (pierreguilmin) [reply dynamiqs/time_array.py:215]: Also in `dynamiqs/utils/quantum_utils/general.py`.

File: dynamiqs/qarrays/utils.py [hunk 2]
```diff
@@ -80,3 +179,220 @@ def stack(qarrays: Sequence[QArray], axis: int = 0) -> QArray:
...
+        QArray: shape=(3, 2, 2), dims=(2,), dtype=float32, layout=dia, ndiags=1
+        [[[1.+0.j   ⋅   ]
+          [  ⋅    1.+0.j]]
+        <BLANKLINE>
+         [[1.+0.j   ⋅   ]
+          [  ⋅    1.+0.j]]
+        <BLANKLINE>
+         [[1.+0.j   ⋅   ]
+          [  ⋅    1.+0.j]]]
+    """
```
Comment (pierreguilmin) [inline dynamiqs/qarrays/utils.py:356]: There's something fishy with the indicated and actual types here.

**Review by pierreguilmin — CHANGES_REQUESTED** [review]
Huge work 🤯 Minor comments + some commits (feel free to revert anything). I also added a doctest that fails, but I think it's a bug, it should pass and is a nice example for the documentation. 😇

File: dynamiqs/time_array.py (hunk 1 above)
Comment (gautierronan) [reply dynamiqs/time_array.py:215]: What do you find annoying? I think a QArray should always be complex valued, so IMO it's expected. We don't have to enforce this dtype (easy to change), but then the dtype can be `int32` or whatever else.

File: dynamiqs/qarrays/utils.py (hunk 2 above)
Comment (gautierronan) [reply dynamiqs/qarrays/utils.py:359]: It's related to the question above, on whether we want to enforce the dtype to be complex or not. For now, `sparsedia_from_dict` is NOT enforcing the dtype, but `to_jax()` (called by `SparseDIAQArray.__repr__`) is, hence the mismatch.

**Comment by pierreguilmin** [comment]
I've added the dtype issue to the meta linear ticket on qarray, merging this as soon as the CI passes.

**Review by pierreguilmin — APPROVED** [review]

**Pull request merged**